drivers/adc/adc_stm32.c

```diff
@@ -758,71 +772,100 @@ static int start_read(const struct device *dev,
 		return -EINVAL;
 	}
 
-	uint32_t channels = sequence->channels;
-	uint8_t index = find_lsb_set(channels) - 1;
+	data->buffer = sequence->buffer;
+	data->channels = sequence->channels;
+	data->channel_count = POPCOUNT(data->channels);
+	data->samples_count = 0;
 
-	if (channels > BIT(index)) {
-		LOG_ERR("Only single channel supported");
-		return -ENOTSUP;
+	if (data->channel_count == 0) {
+		LOG_ERR("No channels selected");
+		return -EINVAL;
 	}
 
-	adc_stm32_setup_channels(dev, index);
+	if (data->channels > BIT(STM32_CHANNEL_COUNT) - 1) {
+		LOG_ERR("Channels bitmask uses out of range channel");
+		return -EINVAL;
+	}
 
-	data->buffer = sequence->buffer;
+#if !defined(CONFIG_SOC_SERIES_STM32F0X) && \
+	!defined(CONFIG_SOC_SERIES_STM32G0X) && \
+	!defined(CONFIG_SOC_SERIES_STM32L0X) && \
+	!defined(CONFIG_SOC_SERIES_STM32WLX)
+	if (data->channel_count > ARRAY_SIZE(table_seq_len)) {
+		LOG_ERR("Too many channels for sequencer. Max: %d", ARRAY_SIZE(table_seq_len));
+		return -EINVAL;
+	}
+#else
+	if (data->channel_count > 1) {
+		LOG_ERR("This device only supports single channel sampling");
+		return -EINVAL;
+	}
+#endif
+
+	uint8_t channel_id;
+	uint8_t channel_index = 0;
+
+	for (uint32_t channels = data->channels; channels;
+		      channels &= ~BIT(channel_id), channel_index++) {
```

Comment:
Let's put a comment here, as this is part not obvious.

Reply:
Something like this?
``` C
	/* Iterate over selected channels in bitmask keeping track of:
	 * - channel_index: ranging from 0 -> ( data->channel_count - 1 )
	 * - channel_id: ordinal position of channel in data->channels bitmask
	 */
	for (uint32_t channels = data->channels; channels;
```